## Gathering ressources & data 

#### Scraping all topics first 

In [18]:
import requests
from bs4 import BeautifulSoup
import re

def scrape_ted_topics():
    """
    Scrapes all topics from the TED topics page using a more robust selector.
    """
    # The URL of the page to scrape
    url = "https://www.ted.com/topics"

    try:
        # Send an HTTP GET request to the URL
        response = requests.get(url)
        # Raise an exception for bad status codes (4xx or 5xx) 
        response.raise_for_status()

        # Parse the HTML content of the page with BeautifulSoup
        soup = BeautifulSoup(response.text, 'html.parser')

        # Find all <a> tags whose 'href' attribute starts with "/topics/"
        # This is more stable than relying on a CSS class.
        topic_elements = soup.select('a[href^="/topics/"]')

        # Extract the text from each element. We filter out any empty strings.
        topics = [element.get_text(strip=True) for element in topic_elements if element.get_text(strip=True)]

        # Print all the scraped topics
        if topics:
            print("Successfully scraped TED topics:")
            for topic in sorted(topics): # Sorting the list alphabetically
                print(f"- {topic}")
        else:
            print("No topics found. The page structure might have changed again.")

    except requests.exceptions.RequestException as e:
        print(f"An error occurred during the HTTP request: {e}")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")

if __name__ == "__main__":
    scrape_ted_topics()

Successfully scraped TED topics:
- 3D printing
- AI
- AIDS
- Activism
- Adaptation
- Addiction
- Africa
- Aging
- Agriculture
- Algorithm
- Aliens
- Alzheimer's
- Ancient world
- Animals
- Animation
- Antarctica
- Anthropocene
- Anthropology
- Archaeology
- Architecture
- Art
- Artificial intelligence
- Asia
- Asteroid
- Astrobiology
- Astronomy
- Atheism
- Augmented reality
- Autism spectrum disorder
- Bacteria
- Beauty
- Bees
- Behavioral economics
- Best of the Web
- Big Bang
- Biodiversity
- Bioethics
- Biology
- Biomimicry
- Bionics
- Biosphere
- Biotech
- Birds
- Blindness
- Blockchain
- Body language
- Books
- Botany
- Brain
- Brazil
- Buddhism
- Bullying
- Business
- CRISPR
- Cancer
- Capitalism
- Career
- Chemistry
- China
- Christianity
- Cities
- Climate change
- Code
- Cognitive science
- Collaboration
- Comedy
- Communication
- Community
- Compassion
- Competition
- Computers
- Conducting
- Confidence
- Consciousness
- Conservation
- Consumerism
- Content Curation
- Coral 

In [19]:
import requests
from bs4 import BeautifulSoup
import csv

def scrape_and_save_ted_topics():
    """
    Scrapes all topics from the TED topics page and saves them to a CSV file.
    """
    url = "https://www.ted.com/topics"
    # Define the name of the output CSV file
    filename = "ted_topics.csv"

    try:
        # 1. Scrape the webpage
        response = requests.get(url)
        response.raise_for_status()  # Check for request errors

        soup = BeautifulSoup(response.text, 'html.parser')
        topic_elements = soup.select('a[href^="/topics/"]')
        topics = [element.get_text(strip=True) for element in topic_elements if element.get_text(strip=True)]

        if not topics:
            print("No topics found. The page structure might have changed.")
            return

        # 2. Write the data to a CSV file
        with open(filename, 'w', newline='', encoding='utf-8') as csvfile:
            # Create a CSV writer object
            writer = csv.writer(csvfile)

            # Write the header row
            writer.writerow(['Topic'])

            # Write the topic data, one topic per row
            for topic in sorted(topics):
                writer.writerow([topic])

        print(f"Scraped {len(topics)} topics and saved them to '{filename}'.")

    except requests.exceptions.RequestException as e:
        print(f"An error occurred during the HTTP request: {e}")
    except IOError as e:
        print(f"An error occurred while writing the file: {e}")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")

if __name__ == "__main__":
    scrape_and_save_ted_topics()

Scraped 383 topics and saved them to 'ted_topics.csv'.


In [20]:
import re
from html import unescape
import pandas as pd

#### For each raw .rtf file, read them and use unescape to match urls and necessary information and get rid of the rest or the raw html data 

In [21]:
# Extract video information using regex patterns
videos = []

# Pattern to find video links with titles, speakers, and duration
pattern = r'href="(/talks/[^"]+)"[^>]*>.*?alt="([^"]*)".*?dir="ltr">(\d+:\d+)</span>.*?dir="ltr">([^<]+)</span>.*?dir="ltr">([^<]+)</p>'
with open('data/collected_raw/12-18.rtf', 'r', encoding='utf-8') as f:
    content = f.read()
matches = re.findall(pattern, content, re.DOTALL) #NAME CONTENT NOT DEFINED

In [22]:
# Extract video information using regex patterns
videos = []

# Pattern to find video links with titles, speakers, and duration
pattern = r'href="(/talks/[^"]+)"[^>]*>.*?alt="([^"]*)".*?dir="ltr">(\d+:\d+)</span>.*?dir="ltr">([^<]+)</span>.*?dir="ltr">([^<]+)</p>'

matches = re.findall(pattern, content, re.DOTALL)

for match in matches:
    url_path = match[0]
    alt_text = unescape(match[1])
    duration = match[2]
    title = unescape(match[3])
    speaker = unescape(match[4])
    
    video_data = {
        'title': title.strip(),
        'speaker': speaker.strip(),
        'duration': duration,
        'url': f'https://ted.com{url_path}'
    }
    videos.append(video_data)

#### Create a dataframe for each and load into a unique .csv file

In [23]:

df = pd.DataFrame(videos)

print(f"Found {len(df)} TED Talk videos\n")
print("DataFrame Shape:", df.shape)
print("\nFirst 5 rows:")
df.head()

print("\nDataFrame Info:")
df.info()

df.to_csv('ted_talks12-18.csv', index=False, encoding='utf-8')
print("\n✓ Data saved to ted_talks12-18.csv")


Found 2265 TED Talk videos

DataFrame Shape: (2265, 4)

First 5 rows:

DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2265 entries, 0 to 2264
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   title     2265 non-null   object
 1   speaker   2265 non-null   object
 2   duration  2265 non-null   object
 3   url       2265 non-null   object
dtypes: object(4)
memory usage: 70.9+ KB

✓ Data saved to ted_talks12-18.csv


In [24]:
df

,title,speaker,duration,url
0,A pastry chef works his chocolatier magic \'97...,Amaury Guichon,12:33,https://ted.com/talks/amaury_guichon_a_pastry_...
1,The flourishing future of women's sports,Kate Johnson,12:48,https://ted.com/talks/kate_johnson_the_flouris...
2,"How we\'92re turning pollution into toys, toot...",Xu Hao,12:53,https://ted.com/talks/xu_hao_how_we_re_turning...
3,The best thing that could happen to the energy...,Matt Tilleard,12:41,https://ted.com/talks/matt_tilleard_the_best_t...
4,3 simple ways to build stronger relationships ...,Alyssa Birnbaum,14:59,https://ted.com/talks/alyssa_birnbaum_3_simple...
...,...,...,...,...
2260,Meet the founder of the blog revolution,Mena Trott,16:29,https://ted.com/talks/mena_trott_meet_the_foun...
2261,Simple designs to save a life,Amy Smith,14:42,https://ted.com/talks/amy_smith_simple_designs...
2262,One Laptop per Child,Nicholas Negroponte,17:20,https://ted.com/talks/nicholas_negroponte_one_...
2263,Letting go of God,Julia Sweeney,16:15,https://ted.com/talks/julia_sweeney_letting_go...


In [25]:
import glob
import os

#### Process data: combine all different collected videos into one big db

In [39]:
csv_files = glob.glob('data/collected_csv/ted_talks*.csv')

In [40]:
csv_files

['data/collected_csv\\ted_talks0-6.csv',
 'data/collected_csv\\ted_talks12-18.csv',
 'data/collected_csv\\ted_talks18+.csv',
 'data/collected_csv\\ted_talks6-12.csv']

In [52]:
import pandas as pd
import glob
import os
import numpy as np

# Read and combine all CSV files
dfs = [pd.read_csv(f) for f in csv_files]
combined_df = pd.concat(dfs, ignore_index=True)

# Remove duplicates 
initial_count = len(combined_df)
combined_df = combined_df.drop_duplicates(subset=["url"])
final_count = len(combined_df)

print(f" {initial_count - final_count} duplicates removed")
print(f"Final count: {final_count}")

# Shuffling dataframe
np.random.seed(42)
combined_shuffled = combined_df.sample(frac=1, random_state=42).reset_index(drop=True)

# Dividing in two dfs for analysis and model testing 
split_index = len(combined_shuffled) // 2
df_analysis = combined_shuffled.iloc[:split_index]
df_test = combined_shuffled.iloc[split_index:]

# Save dfs in new CSV files 
df_analysis.to_csv("data/ted_talks_analysis.csv", index=False)
df_test.to_csv("data/ted_talks_test.csv", index=False)

# Display the shape and first few rows to confirm
print(f"Combined {len(csv_files)} files.")
print(df_analysis.shape)
print(df_test.shape)

# Checking there is no duplicates
analysis_urls = set(df_analysis['url'])
test_urls = set(df_test['url'])
common_urls = analysis_urls.intersection(test_urls)

print(f"URLs found in commun : {len(common_urls)}")

df_analysis.head()



 23 duplicates removed
Final count: 7244
Combined 4 files.
(3622, 4)
(3622, 4)
URLs found in commun : 0


,title,speaker,duration,url
0,Owning the kitchen again,Einav Gefen,07:56,https://ted.com/talks/einav_gefen_owning_the_k...
1,Myths and misconceptions about evolution,Alex Gendler,04:07,https://ted.com/talks/alex_gendler_myths_and_m...
2,The Web's secret stories,Jonathan Harris,16:53,https://ted.com/talks/jonathan_harris_the_web_...
3,How disgust drives your politics,Cindy Kam,15:50,https://ted.com/talks/cindy_kam_how_disgust_dr...
4,How we found the worst place to park in New Yo...,Ben Wellington,11:34,https://ted.com/talks/ben_wellington_how_we_fo...


In [48]:
print("hello")
df_test.head()

hello


,title,speaker,duration,url
3622,I am the son of a terrorist. Here's how I chos...,Zak Ebrahim,08:56,https://ted.com/talks/zak_ebrahim_i_am_the_son...
3623,Why don\'92t we get our drinking water from th...,Manish Kumar,05:10,https://ted.com/talks/manish_kumar_why_don_t_w...
3624,"Visions of Africa's future, from African filmm...",Dayo Ogunyemi,11:32,https://ted.com/talks/dayo_ogunyemi_visions_of...
3625,The psychology of evil,Philip Zimbardo,22:53,https://ted.com/talks/philip_zimbardo_the_psyc...
3626,Let's rethink America's military strategy,Thomas Barnett,23:24,https://ted.com/talks/thomas_barnett_let_s_ret...


#### In total the site contains (precise date of retrieval, as new videos are available on the website)
* 1935 0-6 min videos
* 2128 6-12 min videos
* 2265 12-18 min videos
* 985 18+ min videos

#### A total of: 7313 videos 
#### We've been able to scrape 99% of the website successfully 
